synapse/datamodel.py

```diff
@@ -109,8 +109,11 @@ def __init__(self, modl, form, name, typedef, info):
             self.isext = name.startswith('._')
         self.isform = False     # for quick Prop()/Form() detection
 
-        self.setperm = ('node', 'prop', 'set', self.full)
         self.delperm = ('node', 'prop', 'del', self.full)
+        self.setperms = [('node', 'prop', 'set', self.full)]
+
+        if form is not None:
+            self.setperms.append(('node', 'prop', 'set', form.name, self.name))
```

Comment:
reverse the list after appending this to it. avoids having to do a list reversal at runtime every time we do the property set.

Reply:
not doing any list reversal at eval time, just checking allowed on `[:-1]` and then confirm on `[-1]` so that the perm it complains about being missing is the most modern one :)